File: java/spring-boot-rabbitmq/rabbitmq-domain/src/test/java/com/meridal/examples/springbootrabbitmq/domain/RecordingFactoryTest.java
package com.meridal.examples.springbootrabbitmq.domain;

import org.junit.Test;

import java.util.ArrayList;
import java.util.List;

import static org.junit.Assert.*;

public class RecordingFactoryTest {

    private static final String ARTIST = "Anne Artist";
    private static final String ID = "RECORDING-1";
    private static final String TITLE = "A Recording Title";
    private static final String CAT_NO = "CAT001";
    private static final Integer YEAR = 1965;
    private static final String SONG_TITLE = "A Song Title";
    private static final int MINS = 5;
    private static final int SECS = 17;
    private static final String DURATION = "05:17";
    private static final String OTHER_ID = "RECORDING-2";

    @Test
    public void testCreateSong() {
        final Song song = RecordingFactory.createSong(SONG_TITLE, MINS, SECS);
        this.validateSong(song);
    }

    @Test
    public void testRandomSong() {
        final Song song = RecordingFactory.randomSong();
        this.validateRandomSong(song);
    }

    @Test
    public void testCreateRecording() {
        final Recording recording = RecordingFactory.createRecording(ID, ARTIST, TITLE, CAT_NO, YEAR);
        this.validateRecording(recording, ID, 0);
    }

    @Test
    public void testCreateRecordingWithSongs() {
        final Song song = RecordingFactory.createSong(SONG_TITLE, MINS, SECS);
        final List<Song> songs = new ArrayList<>();
        songs.add(song);
        final Recording recording = RecordingFactory.createRecording(ID, ARTIST, TITLE, CAT_NO, YEAR, songs);
        this.validateRecording(recording, ID, 1);
        this.validateSong(recording.getSongs().get(0));
    }

    @Test
    public void testCloneRecording() {
        final Song song = RecordingFactory.createSong(SONG_TITLE, MINS, SECS);
        final List<Song> songs = new ArrayList<>();
        songs.add(song);
        final Recording recording = RecordingFactory.createRecording(ID, ARTIST, TITLE, CAT_NO, YEAR, songs);
        final Recording other = RecordingFactory.cloneRecording(OTHER_ID, recording);
        this.validateRecording(other, OTHER_ID, 1);
        this.validateSong(other.getSongs().get(0));
    }

    @Test
    public void testRandomRecordingWithID() {
        final Recording recording = RecordingFactory.randomRecording(ID);
        this.validateRandomRecording(recording, ID);
    }

    @Test
    public void testRandomRecordingWithNoID() {
        final Recording recording = RecordingFactory.randomRecording();
        this.validateRandomRecording(recording, null);
    }

    private void validateSong(final Song song) {
        assertNotNull(song);
        assertEquals(SONG_TITLE, song.getTitle());
        assertEquals(DURATION, song.getDuration());
    }

    private void validateRecording(final Recording recording, final String id, final int numSongs) {
        assertNotNull(recording);
        assertNotNull(recording.getSongs());
        assertEquals(numSongs, recording.getSongs().size());
        assertEquals(id, recording.getId());
        assertEquals(ARTIST, recording.getArtist());
        assertEquals(TITLE, recording.getTitle());
        assertEquals(CAT_NO, recording.getCatalogueNumber());
        assertEquals(YEAR, recording.getYear());
    }

    private void validateRandomSong(final Song song) {
        assertNotNull(song);
        assertNotNull(song.getTitle());
        assertNotNull(song.getDuration());
    }

    private void validateRandomRecording(final Recording recording, final String id) {
        assertNotNull(recording);
        assertNotNull(recording.getSongs());
        assertTrue(recording.getSongs().size() > 0);
        assertEquals(id, recording.getId());
        assertNotNull(recording.getArtist());
        assertNotNull(recording.getTitle());
        assertNotNull(recording.getCatalogueNumber());
        assertNotNull(recording.getYear());
    }
}

File: kotlin/kotlin-domain/src/test/kotlin/com/meridal/examples/domain/RecordingFactoryTest.kt
package com.meridal.examples.domain

import org.junit.Assert.*
import org.junit.Test

class RecordingFactoryTest {

    private val ARTIST = "Anne Artist"
    private val ID = "RECORDING-1"
    private val TITLE = "A Recording Title"
    private val CAT_NO = "CAT001"
    private val YEAR = 1965
    private val SONG_TITLE = "A Song Title"
    private val MINS = 5
    private val SECS = 17
    private val DURATION = "05:17"
    private val OTHER_ID = "RECORDING-2"

    @Test
    fun testCreateSong() {
        val song: Song = createSong(SONG_TITLE, MINS, SECS)
        validateSong(song)
    }

    @Test
    fun testRandomSong() {
        val song: Song = randomSong()
        validateRandomSong(song)
    }

    @Test
    fun testCreateRecordingWithSongs() {
        val song: Song = createSong(SONG_TITLE, MINS, SECS)
        val songs: MutableList<Song> = mutableListOf()
        songs.add(song)
        val recording: Recording = createRecording(ID, ARTIST, TITLE, CAT_NO, YEAR, songs)
        validateRecording(recording, ID, 1)
        validateSong(recording.songs[0])
    }

    @Test
    fun testCloneRecording() {
        val song: Song = createSong(SONG_TITLE, MINS, SECS)
        val songs: MutableList<Song> = mutableListOf()
        songs.add(song)
        val recording: Recording = createRecording(ID, ARTIST, TITLE, CAT_NO, YEAR, songs)
        val other: Recording = cloneRecording(OTHER_ID, recording)
        validateRecording(other, OTHER_ID, 1)
        validateSong(other.songs[0])
    }

    @Test
    fun testRandomRecordingWithID() {
        val recording: Recording = randomRecording(ID)
        validateRandomRecording(recording, ID)
    }

    private fun validateRecording(recording: Recording, id: String, numSongs: Int) {
        assertNotNull(recording.songs)
        assertEquals(numSongs, recording.songs.size)
        assertEquals(id, recording.id)
        assertEquals(ARTIST, recording.artist)
        assertEquals(TITLE, recording.title)
        assertEquals(CAT_NO, recording.catalogueNumber)
        assertEquals(YEAR, recording.year)
    }

    private fun validateRandomRecording(recording: Recording, id: String) {
        assertNotNull(recording.songs)
        assertTrue(recording.songs.size > 0)
        assertEquals(id, recording.id)
        assertNotNull(recording.artist)
        assertNotNull(recording.title)
        assertNotNull(recording.catalogueNumber)
        assertNotNull(recording.year)
    }

    private fun validateSong(song: Song) {
        assertEquals(SONG_TITLE, song.title)
        assertEquals(DURATION, song.duration)
    }

    private fun validateRandomSong(song: Song) {
        assertNotNull(song.title)
        assertNotNull(song.duration)
    }
}